src/Generics/Deriving/Default.hs

```diff
@@ -0,0 +1,276 @@
+-- Stability   : experimental
+-- Portability : non-portable
+--
+-- GHC 8.6 introduced the 'DerivingVia' language extension, which means
+-- a typeclass instance can be derived from an existing instance for an
+-- isomorphic type. Any newtype is isomorphic to the underlying type. By
+-- implementing a typeclass once for the newtype, it is possible to derive
+-- any typeclass for any type with a 'Generic' instance.
+--
+-- For a number of classes, there are sensible default instantiations. In
+-- older GHCs, these can be supplied in the class definition, using the
+-- 'DefaultSignatures' extension. However, only one default can be
+-- provided! With 'DerivingVia' it is now possible to choose from many
+-- default instantiations.
+--
+-- This package contains a number of such classes. This module demonstrates
+-- how one might create a newtype called 'Default' for which such instances
+-- are defined.
+--
+-- Then one might use 'DerivingVia' as follows. The implementations of the
+-- data types are elided here (they are irrelevant). Either the deriving
+-- clause with the data type definition or the standalone clause will work.
+--
+-- 1.  For data types of kind '*', use 'Default'.
+--
+-- @
+-- data MyType = …
+--  deriving (Generic)
+--  deriving (GEq) via (Default MyType)
+--
+-- deriving via (Default MyType) instance GShow MyType
+-- @
+--
+-- TODO Confirm classes of kind '* -> Constraint' vs kind 'Constraint'
+--
+-- 2.  For data types of kind '* -> *', use 'Default1'.
+--
+-- @
+-- data MyType1 a = …
+--  deriving (Generic)
+--  deriving (GFunctor) via (Default1 MyType1)
+--
+-- deriving via (Default1 MyType1) instance GFoldable MyType1
+-- @
+
+{-# LANGUAGE CPP #-}
+#if __GLASGOW_HASKELL__ >= 701
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE Safe #-}
+#endif
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Generics.Deriving.Default
+  ( Default(..)
+  , Default1(..)
+  ) where
+
+import Generics.Deriving.Base
+import Generics.Deriving.Copoint
+import Generics.Deriving.Enum
+import Generics.Deriving.Eq
+import Generics.Deriving.Foldable
+import Generics.Deriving.Functor
+import Generics.Deriving.Monoid
+import Generics.Deriving.Semigroup
+import Generics.Deriving.Show
+import Generics.Deriving.Traversable
+import Generics.Deriving.Uniplate
+
+-- | This newtype wrapper can be used to derive default instances for data
+-- types of kind '*'.
+newtype Default a = Default { unDefault :: a }
+
+-- | This newtype wrapper can be used to derive default instances for data
+-- types of kind '* -> *'.
+newtype Default1 f a = Default1 { unDefault1 :: f a }
+
+--------------------------------------------------------------------------------
+-- Eq
+--------------------------------------------------------------------------------
+
+-- |
+--
+-- @
+-- instance (Generic a, GEq' (Rep a)) => Eq (Default a) where
+--   (==) :: Default a -> Default a -> Bool
+--   Default x == Default y = x `geqdefault` y
+-- @
```

Comment:
There's no need to keep these comments around, right?

Reply:
I was originally thinking that they could be useful as part of the haddock documentation, but I now think they only cause confusion.